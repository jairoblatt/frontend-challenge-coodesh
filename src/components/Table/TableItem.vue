<template>
  <tr class="table-item">
    <td class="item__avatar">
      <Avatar v-if="avatar" size="2.1" border="1.5" class="mr-2">
        <img :src="avatar" :alt="`${nameResolve} avatar`" />
      </Avatar>
      {{ nameResolve }}
    </td>
    <td>{{ gender }}</td>
    <td>{{ age }}</td>
    <td><slot /></td>
  </tr>
</template>

<script lang="ts">
import Vue, { PropOptions, PropType } from "vue";
import { Nullable, UserName } from "@/types/types";

export default Vue.extend({
  props: {
    gender: String as PropType<string>,
    age: Number as PropType<number>,
    name: {
      type: Object,
      default: null,
    } as PropOptions<Nullable<UserName>>,
    avatar: String as PropType<string>,
  },

  computed: {
    nameResolve(): string {
      const { name } = this;
      return name ? `${name.first} ${name.last}` : "Unknow name";
    },
  },
});
</script>

<style lang="postcss" scoped>
.table-item td {
  @apply py-3 px-2 text-light-onSurfacePrimary;
}

.item__avatar {
  @apply flex items-center;
}
</style>
