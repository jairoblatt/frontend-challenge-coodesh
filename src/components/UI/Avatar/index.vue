<template>
  <component :is="tag" class="avatar" :style="styles">
    <slot />
  </component>
</template>

<script lang="ts">
import Vue, { PropOptions, PropType } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Avatar",

  props: {
    tag: {
      type: String,
      default: "div",
    } as PropOptions<string>,
    // The prop size it's definedly with 'rem' css unit meansure
    size: {
      type: [String, Number],
      default: 2,
    } as PropOptions<string | number>,
    // The prop border it's definedly with 'px' css unit meansure
    border: [String, Number] as PropType<string | number>,
  },

  computed: {
    styles(): CSS.Properties {
      return {
        height: `${this.size}rem`,
        width: `${this.size}rem`,
        borderWidth: `${this.border || Number(this.size) / 2}px`,
      };
    },
  },
});
</script>

<style lang="postcss" scoped>
.avatar {
  @apply rounded-full border-4 border-primary-base border-solid overflow-hidden;
}

.avatar img,
source {
  @apply min-w-full h-auto object-contain;
}

/* absolute -top-28 left-2/4 transform -translate-x-2/4 */

/* .user__avatar img,
source {
  height: 12rem;
  width: 12rem;
} */
</style>
