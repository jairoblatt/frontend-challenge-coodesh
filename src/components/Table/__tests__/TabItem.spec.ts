import { mount, Wrapper, MountOptions, createLocalVue } from "@vue/test-utils";
import TableItem from "../TableItem.vue";
import Avatar from "@/components/UI/Avatar/index.vue";

describe("TableItem.vue ", () => {
  const localVue = createLocalVue();
  localVue.component("Avatar", Avatar);

  type Instance = InstanceType<typeof TableItem>;
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>;

  const propsContent = {
    avatar: "something.png",
    gender: "male",
    age: 24,
    name: {
      first: "Wonderful",
      last: "Name",
    },
  };

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(TableItem, {
        ...options,
        localVue,
      });
    };
  });

  it("Should mount the component", () => {
    const wrapper = mountFunction();

    expect(wrapper.vm).toBeDefined();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should render the avatar component when avatar prop is passed", () => {
    const wrapper = mountFunction({
      propsData: propsContent,
    });

    expect(wrapper.findComponent(Avatar).exists()).toBe(true);
  });

  it("Should render the default slot when him is passed", () => {
    const slot = `<h5>Hello</h5>`;

    const wrapper = mountFunction({
      slots: {
        default: slot,
      },
    });

    expect(wrapper.find("h5").text()).toBe("Hello");
  });

  it("Should return the user fully name", () => {
    const wrapper = mountFunction({
      propsData: propsContent,
    });

    expect((wrapper.vm as any).nameResolve).toBe("Wonderful Name");
  });
});
