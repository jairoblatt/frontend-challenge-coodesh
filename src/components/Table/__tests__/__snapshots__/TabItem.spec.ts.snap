// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableItem.vue  Should mount the component 1`] = `
"<tr class=\\"table-item\\">
  <td class=\\"item__avatar\\">
    <!---->
    Unknow name
  </td>
  <td></td>
  <td></td>
  <td></td>
</tr>"
`;
